feat: Exporter business logic for list-view index

mihail-ca opened the pull request:
feat: Exporter business logic for list-view index

- added handlers for list-view index

Closes https://github.com/camunda/camunda/issues/18908

mihail-ca commented:
The handlers are based on the following importer methods, in the importer class ListViewZeebeRecordProcessor.java:

```
ListViewFlowNodeFromIncidentHandler  				--- based on processIncidentRecord()
ListViewFlowNodeFromJobHandler  					--- based on processJobRecords()
ListViewFlowNodeFromProcesInstanceHandler  			--- based on processProcessInstanceRecord()
ListViewProcessInstanceFromProcessInstanceHandler  	--- based on processProcessInstanceRecord()
ListViewVariableFromVariableHandler  				--- based on processVariableRecords()
```

The handler `ListViewProcessInstanceFromProcessInstanceHandler` is problematic, I implemented as much as I could. But as you can see, it has multiple external dependencies, added via constructor. I am not sure the implementation is correct here, also which importer code should have been added to the method `updateEntity()`, and which to the method `flush()`. Also in the importer code, this handler was using a `ProcessCache`, which I am not sure I imported correctly. I did not add this handler to the `OperateElasticsearchExporter.java`, as it has too many dependencies, and the implementation may need rework. 
FYI @sdorokhova @ThorbenLindhauer

sdorokhova commented:
I will document here what we discussed about `ListViewProcessInstanceFromProcessInstanceHandler`:
* `OperationsManager` dependency. We will rework (simplify) how we complete operations in #18372. Therefore for now I propose to comment out the calls to `OperationsManager`. But the code block should be moved to `update` method, so that we don't need to have `recordsMap` local cache.
* same case with `ProcessCache`. We will rework it in #18364. So for now let's comment it out and put `bpmnProcessId` as process definition name.
* same case with `ListViewStore` and `FlowNodeStore`. We use them to build process instance tree path that we will rework in #18378. Let's again remove the corresponding methods and put null as `treePath` for now.
* dependency on `MetricsStore`. We use it to record root process instance creation event. These are business related metrics which we store in Elasticsearch as `MetricEntity`. We decided that since we have an Operate entity for this we should create a separate handler for this.
* dependency on `ImportBatch`. This is used to record Prometheus metric [counting finished process instances](https://github.com/camunda/camunda/blob/ff8dbe135d1523283ba1324ed42c98824150432d/operate/importer/src/main/java/io/camunda/operate/zeebeimport/EventsProcessedMetricsCounterImportListener.java#L36). Here we should research whether it's OK from performance point of view to record finished events one by one? Or we really need to accumulate them and flush only from time to time? In this case we can create a separate issue and address this later.

Every time when we delete/comment out some code, let's put `TODO` with the link to GitHub issue.

sdorokhova commented on zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/handlers/ListViewProcessInstanceFromProcessInstanceHandler.java:
```diff
@@ -0,0 +1,348 @@
+      Record<ProcessInstanceRecordValue> record, ProcessInstanceForListViewEntity entity) {
+
+    final var recordValue = record.getValue();
+    if (isProcessEvent(recordValue)) {
```
This check can be moved to `handlesRecord` method.

mihail-ca commented:
Based on the comments above, I made the following changes to `ListViewProcessInstanceFromProcessInstanceHandler`:

- I removed from constructor of `ListViewProcessInstanceFromProcessInstanceHandler` references to the additional parameters: `ImportBatch ListViewStore FlowNodeStore OperationsManager ProcessCache`
- To keep the original logic from the importer, I added dummy methods at the end (marked with TODO), that define the needed references as local null variables, and return a default value, or take no action (since the references are null). I think this makes it easier to keep track of the original logic in the importer, and later we can see how to implement or replace these dummy methods, marked with TODO
- I also created a separate handler for the Operate metrics, similar to other handlers, so I removed the code related to metrics from `ListViewProcessInstanceFromProcessInstanceHandler`

FYI @sdorokhova @ThorbenLindhauer

sdorokhova commented on zeebe/exporters/operate-exporter/pom.xml:
```diff
@@ -24,6 +24,11 @@
 
+    <dependency>
+      <groupId>org.springframework</groupId>
+      <artifactId>spring-core</artifactId>
```
We have this dependency for `StringUtils` class, right? Can we may be extend our own utility classes? Would be nice not to have Spring dependency on this module.

mihail-ca replied:
@sdorokhova I do not use this dependency, but I added it because the CI fails without it, I do not understand why:
<img width="694" alt="image" src="https://github.com/user-attachments/assets/35402f1a-9cb2-4521-a1df-0897ec4db578">

mihail-ca replied:
I will merge like this, to have the CI pass, as this dependency can be removed later.

sdorokhova commented on zeebe/exporters/operate-exporter/pom.xml:
```diff
@@ -84,6 +89,11 @@
 
+    <dependency>
+      <groupId>io.camunda</groupId>
+      <artifactId>operate-importer-common</artifactId>
```
Similar question about this dependency. Why do we need it? I think it's OK to keep it for now, but would be nice to create an issue to clean this up later.

sdorokhova reviewed (approved):
I left couple of minor comments on pom, but other than that changes look good!

I will approve the PR, so that you don't need to wait for another review, but please check my comments.
